log4j-core/src/main/java/org/apache/logging/log4j/core/appender/routing/RoutingAppender.java

```diff
@@ -224,13 +224,16 @@ public void append(LogEvent event) {
             control.callAppender(event);
         }
 
-        if (purgePolicy != null) {
+        if (purgePolicy != null
+                // LOG4J2-2631: PurgePolicy implementations do not need to be aware of appenders that
+                // were not created by this RoutingAppender.
+                && !referencedAppenders.containsKey(key)) {
```

Comment:
This check is a bit odd, I considered another approach where the concurrent map uses a value with additional metadata describing whether or not the appender is based on a reference. Unfortunately that approach would still require two concurrent maps in order to avoid concurrency issues, so I opted against it in favor of the containsKey check here.